File: love2d/src/sparks.lua
-- Small bursts, thrown at a point on the screen.
--
-- ANSWER mode needs something to happen where a mistake is, and the whole of
-- what it needs is: a dozen squares thrown outwards, easing to a stop and then
-- falling, fading as they go. The browser client has a particle engine and a
-- confetti planner for its result screen; this is the same idea at the size
-- this client actually uses it, in one file, with no dependencies beyond the
-- clock the scene already keeps.
--
-- **A particle's position is a function of its age**, not something integrated
-- between frames. That is what makes a burst look the same on a 60 Hz panel
-- and a 144 Hz one, and it is why `update` only advances a clock.

local Sparks = {}
Sparks.__index = Sparks

--- How long the longest particle lives. Nothing is kept past this.
local LIFE = 1.25

function Sparks.new()
  return setmetatable({ t = 0, bursts = {} }, Sparks)
end

--- `n` squares in `color`, from `(x, y)`.
---
--- `n` is the whole dial: ten is "you typed a character that is not the
--- answer", seventy is "you have typed the entire thing, exactly".
function Sparks:add(x, y, n, color)
  local parts = {}
  for i = 1, n do
    local angle = math.random() * math.pi * 2
    local reach = 40 + math.random() * 110 * math.sqrt(math.max(0.6, n / 24))
    parts[i] = {
      x = x,
      y = y,
      dx = math.cos(angle) * reach,
      dy = math.sin(angle) * reach - 30,
      size = 2 + math.random() * 4,
      life = 0.5 + math.random() * (LIFE - 0.5),
      delay = math.random() * 0.05,
      gravity = 150 + math.random() * 120,
    }
  end
  self.bursts[#self.bursts + 1] = {
    born = self.t,
    color = color,
    parts = parts,
    -- The ring is what makes a burst read as *one event* rather than as a
    -- handful of dots that happen to have started together.
    ring = 10 + n * 0.6,
  }
  -- A cap, not a queue: a player typing fast can ask for one per keystroke,
  -- and what matters on screen is the last few.
  if #self.bursts > 8 then table.remove(self.bursts, 1) end
end

function Sparks:update(dt)
  self.t = self.t + dt
  -- Drop what has finished, so an idle screen costs nothing.
  for i = #self.bursts, 1, -1 do
    if self.t - self.bursts[i].born > LIFE + 0.1 then table.remove(self.bursts, i) end
  end
end

function Sparks:clear()
  self.bursts = {}
end

function Sparks:draw()
  for _, b in ipairs(self.bursts) do
    local age = self.t - b.born
    local c = b.color
    if age < 0.45 then
      local u = age / 0.45
      love.graphics.setColor(c[1], c[2], c[3], (1 - u) * 0.55)
      love.graphics.setLineWidth(math.max(1, 3 * (1 - u)))
      love.graphics.circle("line", b.parts[1] and b.parts[1].x or 0,
        b.parts[1] and b.parts[1].y or 0, b.ring * (1 - (1 - u) * (1 - u)), 24)
      love.graphics.setLineWidth(1)
    end
    for _, p in ipairs(b.parts) do
      local a = age - p.delay
      if a > 0 and a < p.life then
        local u = a / p.life
        -- Ease out along the throw, then gravity on top of it.
        local out = 1 - (1 - u) * (1 - u) * (1 - u)
        local x = p.x + p.dx * out
        local y = p.y + p.dy * out + 0.5 * p.gravity * a * a
        love.graphics.setColor(c[1], c[2], c[3], math.max(0, 1 - u * u))
        love.graphics.rectangle("fill", x - p.size / 2, y - p.size / 2, p.size, p.size)
      end
    end
  end
  love.graphics.setColor(1, 1, 1, 1)
end

return Sparks

File: frontend/src/ui/sparks.ts
/**
 * A spark layer that sits *over* the DOM editor.
 *
 * The game has two canvases and both of them are underneath `#overlay`
 * (`index.html` says so in as many words), and the editor's own face is 96%
 * opaque — so an effect thrown at the caret and painted on the game canvas is
 * an effect nobody sees. ANSWER mode's whole point is to put something where
 * the mistake is, so this is a third, transparent canvas, added *inside* the
 * overlay after the editor and therefore painted on top of it.
 *
 * It takes the same virtual coordinates the rest of the game draws in, and
 * sets the same transform `Layout.begin` does, so a burst asked for at a point
 * in the scene lands on that point. It never takes a pointer event — it is
 * decoration over something a person is typing into, and a decoration that
 * swallowed a click would be a bug in the editor.
 */
import { cosine, expOut } from "../engine/ease";
import { pathPoint, type Plan, type Smear } from "../engine/burst";
import type { Layout } from "../engine/layout";
import { css } from "../engine/theme";

export class Sparks {
  private readonly canvas = document.createElement("canvas");
  private readonly g: CanvasRenderingContext2D | null;
  /** Bursts, with the scene clock reading at which each was thrown. */
  private live: Array<{ plan: Plan; born: number }> = [];
  /** Caret smears, the same way. */
  private smears: Array<{ smear: Smear; born: number }> = [];

  constructor(
    host: HTMLElement,
    private readonly layout: Layout,
  ) {
    this.canvas.className = "cwb-sparks";
    this.g = this.canvas.getContext("2d");
    host.appendChild(this.canvas);
  }

  /** Throw one. `now` is the scene's clock, in seconds. */
  add(plan: Plan, now: number): void {
    this.live.push({ plan, born: now });
    // A cap rather than a queue: a player typing fast can ask for one a
    // keystroke, and what matters is the last few.
    if (this.live.length > 8) this.live.shift();
  }

  /** The caret moved: its smear, as a stroked path that shortens from the tail. */
  smear(s: Smear, now: number): void {
    this.smears.push({ smear: s, born: now });
    if (this.smears.length > 6) this.smears.shift();
  }

  clear(): void {
    this.live.length = 0;
    this.smears.length = 0;
    this.paintNothing();
  }

  /**
   * Paint whatever is still alive at `now`.
   *
   * The path is closed-form — position is a function of age, exactly as the
   * result screen's confetti — so it looks the same at 60 Hz and at 144 Hz
   * and nothing is integrated between frames.
   */
  draw(now: number): void {
    const g = this.g;
    if (!g) return;
    if (this.live.length === 0 && this.smears.length === 0) return;
    const { dw, dh, scale, ox, oy } = this.layout;
    if (this.canvas.width !== dw || this.canvas.height !== dh) {
      this.canvas.width = dw;
      this.canvas.height = dh;
    }
    g.setTransform(1, 0, 0, 1, 0, 0);
    g.clearRect(0, 0, dw, dh);
    g.setTransform(scale, 0, 0, scale, ox, oy);

    let alive = false;
    for (const { smear, born } of this.smears) {
      const k = (now - born) / smear.life;
      if (k < 0 || k >= 1) continue;
      alive = true;
      // The same shape as the GPU's: the head across almost at once, the
      // tail on a slow start, width and alpha dying together.
      const head = expOut(Math.min(1, k / 0.45));
      const tail = k * k * (3 - 2 * k);
      if (head <= tail) continue;
      const piece = (from: number, to: number): [number, number][] => {
        const pts: [number, number][] = [pathPoint(smear.path, from)];
        if (smear.path.length === 3) {
          const [ax, ay] = smear.path[0];
          const [cx, cy] = smear.path[1];
          const [bx, by] = smear.path[2];
          const first = Math.hypot(cx - ax, cy - ay);
          const corner = first / (first + Math.hypot(bx - cx, by - cy));
          if (corner > from && corner < to) pts.push([cx, cy]);
        }
        pts.push(pathPoint(smear.path, to));
        return pts;
      };
      const pts = piece(tail, head);
      const stroke = (width: number, colour: string, alpha: number) => {
        g.globalAlpha = alpha;
        g.strokeStyle = colour;
        g.lineWidth = Math.max(1, width);
        g.lineCap = "butt";
        g.lineJoin = "miter";
        g.beginPath();
        g.moveTo(pts[0][0], pts[0][1]);
        for (let i = 1; i < pts.length; i++) g.lineTo(pts[i][0], pts[i][1]);
        g.stroke();
      };
      const fade = 1 - k * k;
      stroke(smear.width * (1 - 0.8 * k * k), css(smear.color), 0.45 * fade);
      stroke(Math.max(1, smear.width * 0.12), "#ffffff", smear.core * 0.8 * fade);
      g.globalAlpha = 1;
    }
    for (const { plan, born } of this.live) {
      const t = now - born;
      for (const r of plan.rings) {
        const age = t - r.delay;
        if (age < 0 || age > r.life) continue;
        alive = true;
        const u = age / r.life;
        g.globalAlpha = (1 - u) * (r.glow ? 0.3 : 0.7);
        g.strokeStyle = css(r.color);
        g.lineWidth = Math.max(1, r.radius * 0.06 * (1 - u));
        g.beginPath();
        g.arc(r.x, r.y, r.radius * expOut(u), 0, Math.PI * 2);
        g.stroke();
      }
      g.globalAlpha = 1;
      g.lineWidth = 1;
      for (const p of plan.particles) {
        const age = t - p.delay;
        if (age < 0 || age > p.life) continue;
        alive = true;
        const u = age / p.life;
        const out = expOut(u);
        const leg = cosine(u);
        const lift = Math.sin(Math.PI * u);
        const x = p.x + p.dx * out + p.tox * leg + p.liftx * lift;
        const y = p.y + p.dy * out + p.toy * leg + p.lifty * lift + 0.5 * p.gravity * age * age;
        g.globalAlpha = Math.max(0, 1 - u * u);
        g.fillStyle = css(p.color, 1);
        // A grain is a flat streak, not a square.
        const h = p.shape === 7 ? p.size * 0.2 : p.size;
        g.fillRect(x - p.size / 2, y - h / 2, p.size, h);
      }
      g.globalAlpha = 1;
    }
    if (!alive) {
      this.live.length = 0;
      this.smears.length = 0;
      this.paintNothing();
    }
  }

  private paintNothing(): void {
    const g = this.g;
    if (!g) return;
    g.setTransform(1, 0, 0, 1, 0, 0);
    g.clearRect(0, 0, this.canvas.width, this.canvas.height);
  }

  destroy(): void {
    this.canvas.remove();
  }
}
